cmd/proxygenerator/interceptor.go

```
@@ -112,31 +112,14 @@ func NewPayloadVisitorInterceptor(options PayloadVisitorInterceptorOptions) (grp
 
 		err := invoker(ctx, method, req, response, cc, opts...)
 		if err != nil {
-			if options.Inbound != nil {
-				stat, ok := status.FromError(err)
-				if ok {
-					// user provided payloads can sometimes end up in the status details of
-					// gRPC errors, make sure to visit those as well
-					newStatus := status.New(stat.Code(), stat.Message())
-					var newDetails []protoadapt.MessageV1
-					for _, detail := range stat.Details() {
-						detailAny, ok := detail.(*anypb.Any)
-						if ok && ({{ range $i, $name := .GrpcPayload }}{{ if $i }} || {{ end }}detailAny.MessageName() == "{{$name}}"{{ end }}) {
-							err = VisitPayloads(ctx, detailAny, *options.Inbound)
-							if err != nil {
-								return err
-							}
-						}
-						newDetails = append(newDetails, detailAny)
-					}
-					newStatus, err = newStatus.WithDetails(newDetails...)
-					if err != nil {
-						return err
-					}
-					return newStatus.Err()
-				}
+			if options.Inbound == nil {
+				return err
+			}
+			if s, ok := status.FromError(err); ok {
+				// user provided payloads can sometimes end up in the status details of
+				// gRPC errors, make sure to visit those as well
+				return visitGrpcErrorPayload(ctx, err, s, options.Inbound)
 			}
-			return err
```

I think this `return err` is still needed here, I don't think you want to fall through to success stuff if `status.FromError` happens to not be `ok`.

---

Maybe I'm misunderstanding something, https://github.com/temporalio/api-go/pull/213#discussion_r2014094412, I thought the original desire was to not immediately return err here.

Oh but I guess if a `response` and `err` are both non-nil, we want to visit the payload, but we'd still want to return the error, right?

---

But if both `response` and `err` are non-nil, would we return the err, or `return VisitPayloads(ctx, resMsg, *options.Inbound)` from processing the response?